### PR TITLE
Improve compatibility with Hexagon hardware

### PR DESCRIPTION
**Pull request by steven-johnson**
The customization done via BENCHMARK_OS_QURT works just fine with the Hexagon simulator, but on at least some Hexagon hardware, both `qurt_timer_get_ticks()` and `std::chrono::now()` are broken and always return 0. This fixes the former by using the better-supported (and essentially identical `qurt_sysclock_get_hw_ticks()` call, and the latter by reading a 19.2MHz hardware counter (per suggestion from Qualcomm). Local testing seems to indicate these changes are just as robust under the simulator as before.

### CODE REVIEW

**Comment by LebedevRI** [comment]
Can you please explain (in the commit message too)
why this `ChronoClockNow` workaround is being introduced
instead of fixing the underlying bug?

File: src/timers.h
Comment (LebedevRI) [inline src/timers.h:59]: In `ChooseClockType`, do we end up with `std::chrono::high_resolution_clock`,
or `std::chrono::steady_clock`? Does either of them work, or neither works?

Comment (steven-johnson) [reply src/timers.h:59]: We end up with `high_resolution_clock`, but neither of them works.

File: src/timers.h
Comment (LebedevRI) [inline src/timers.h:25]: Could you please wrap this into a clock-like class,
and change `ChooseClockType` to pick it on `BENCHMARK_OS_QURT`?

**Comment by LebedevRI** [comment]
> > Can you please explain (in the commit message too) why this `ChronoClockNow` workaround is being introduced instead of fixing the underlying bug?
> 
> It is not in my power to fix the underlying bug, which (presumably) lies inside the Qualcomm SDK. I'd much prefer it if the implementation of std::chrono that Qualcomm provided actually works.

Makes sense. Please link the upstream bugreport in commit message.

File: src/timers.h
Comment (steven-johnson) [reply src/timers.h:25]: That seems a lot more complex than actually needed, but OK.

**Comment by steven-johnson** [comment]
> > > Can you please explain (in the commit message too) why this `ChronoClockNow` workaround is being introduced instead of fixing the underlying bug?
> > 
> > 
> > It is not in my power to fix the underlying bug, which (presumably) lies inside the Qualcomm SDK. I'd much prefer it if the implementation of std::chrono that Qualcomm provided actually works.
> 
> Makes sense. Please link the upstream bugreport in commit message.

There is no upstream bug report, because I haven't yet gotten Qualcomm to admit that this is in fact a bug -- I'm trying to find the correct contacts at Qualcomm.

File: src/timers.h
Comment (steven-johnson) [reply src/timers.h:25]: Actually, I'm gonna push back on this a bit: the only reason that the `ChooseClockType` type exists is so that we can call `now()` on it in `ChronoClockNow()`, it's literally the only place in the code base that we use it. Trying to create a class that mimics the existence of `std::chrono` is pointless overhead. If I'm missing something subtle, please let me know, but this seems like a useless exercise.

Comment (LebedevRI) [reply src/timers.h:25]: https://godbolt.org/z/T84efY5fG

Comment (steven-johnson) [reply src/timers.h:25]: I'm not disputing what the object code will look like; I'm disputing that constructing a class would be wrong, because `std::chrono` implies a valid date/time, not just a relative time counter, and we can't mimic that with this approach; it means we have to mimic an API designed to do all sorts of arbitrary things with timestamps, rather than one that simply provides a seconds-counter in the form of a double.

Comment (LebedevRI) [reply src/timers.h:25]: [`std::chrono::steady_clock`](https://en.cppreference.com/w/cpp/chrono/steady_clock):
> This clock is not related to wall clock time (for example, it can be time since last reboot), and is most suitable for measuring intervals.

**Comment by steven-johnson** [comment]
PTAL

File: src/timers.h
Comment (LebedevRI) [inline src/timers.h:71]: Nit: extra newline

File: src/timers.h
Comment (LebedevRI) [inline src/timers.h:33]: (You are sure this shouldn't be `rep`?)

**Comment by LebedevRI** [comment]
Cool, looks good to me. It would be good to get some kind of
an acknowledgement from Qualcomm re the bug,
i'm not sure if we should wait for that or not?

**Comment by LebedevRI** [comment]
And one more thought: in the cases where the normal clocks work, what's their frequency, is it better than of this clock?
I.e., should we be always using this clock, or should we detect the bug, and only then use this "fallback" clock?

**Comment by steven-johnson** [comment]
> Cool, looks good to me. It would be good to get some kind of an acknowledgement from Qualcomm re the bug, i'm not sure if we should wait for that or not?

IMHO we should wait for comment

> And one more thought: in the cases where the normal clocks work, what's their frequency, is it better than of this clock? I.e., should we be always using this clock, or should we detect the bug, and only then use this "fallback" clock?

The normal clock frequencies are implementation-defined. As to whether we should detect and fallback, that would add a bit of extra overhead where we probably want to minimize overhead, but I'll defer to QC folk on that.

**Comment by LebedevRI** [comment]
> > Cool, looks good to me. It would be good to get some kind of an acknowledgement from Qualcomm re the bug, i'm not sure if we should wait for that or not?
> 
> IMHO we should wait for comment

Let's wait.

> > And one more thought: in the cases where the normal clocks work, what's their frequency, is it better than of this clock? I.e., should we be always using this clock, or should we detect the bug, and only then use this "fallback" clock?
> 
> The normal clock frequencies are implementation-defined. As to whether we should detect and fallback, that would add a bit of extra overhead where we probably want to minimize overhead, but I'll defer to QC folk on that.

That would need to happen at CMake time, true, but then i'm not sure if it can truly be detected then,
because i suspect it might not happen when the compiled code is run not on the actual hardware,
but emulated on the machine on which it is being compiled.

**Review by LebedevRI — DISMISSED** [review]
LGTM, but let's wait.

**Comment by steven-johnson** [comment]
Update: It's taking much longer than I expected to get any kind of feedback from Qualcomm on this, and my team is getting blocked by this. Can we go ahead and land this, with the understanding that it may be rolled back and/or fixed forward when we eventually do get feedback from QC?

File: src/timers.cc
Comment (LebedevRI) [inline src/timers.cc:130]: Can you please add in-code comment why this uses `qurt_sysclock_get_hw_ticks` instead of `qurt_timer_get_ticks`?

Comment (steven-johnson) [reply src/timers.cc:134]: Done

**Review by LebedevRI — APPROVED** [review]

**Comment by aankit-quic** [comment]
@steven-johnson @LebedevRI  I'm really sorry. I still haven't gotten any response from the team who could answer these concerns. 

The PR looks good to me. The qurt APIs essentially do the same thing as what Steven is doing here. I feel the std::chrono should have worked in the first place and is probably a bug, but the responsible team still have not confirmed anything.

**Comment by LebedevRI** [comment]
> Update: It's taking much longer than I expected to get any kind of feedback from Qualcomm on this, and my team is getting blocked by this.

FWIW, i've very much expected that it would take a month+ before anything happens here.
Sure, let's merge this.

**Comment by LebedevRI** [comment]
@steven-johnson @aankit-quic thank you!

**Pull request merged**
